src/io/ioexternalbus.cpp

```diff
@@ -28,10 +50,217 @@ QString IOExternalBus::description() const {
 }
 
 VInt IOExternalBus::ioRead(AInt offset, unsigned size) {
-    return 0;
+    uint32_t rvalue = 0;
+
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload = htonl(offset);
+        QByteArray dp = QByteArray(static_cast<const char*>((void*)&payload), 4);
+        if (send_cmd(VB_PREAD, 4, &dp) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(cmd_header) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        if (cmd_header.payload_size) {
+            dp = QByteArray(8, 0);
+            if (recv_payload(dp, 8) < 0) {
+                skt_use = 0;
+                return 0;
+            }
+            uint32_t* payload = (uint32_t*)dp.data();
+
+            for (uint32_t i = 0; i < 2; i++) {
+                payload[i] = ntohl(payload[i]);
+            }
+            dprintf("Read addr[%x] = %x \n", payload[0], payload[1]);
+            rvalue = payload[1];
+        } else {
+            skt_use = 0;
+            printf("read error [%lx] (%x) msg=%i pyaload_size=%i !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n", offset, size,
+                   cmd_header.msg_type, cmd_header.payload_size);
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+
+    dprintf("===============>ioRead off=%lx  size=%x  value=%x\n", offset, size, rvalue);
+    return rvalue;
+}
+
+void IOExternalBus::ioWrite(AInt offset, VInt value, unsigned size) {
+    dprintf("===============>ioWrite off=%lx size=%lx  value=%x\n", offset, size, value);
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload[2];
+        payload[0] = htonl(offset);
+        payload[1] = htonl(value);
+        QByteArray dp = QByteArray(static_cast<char*>((void*)payload), 8);
```

Comment:
casts. For this style of cast, you can use [reinterpret cast](https://en.cppreference.com/w/cpp/language/reinterpret_cast).

Reply:
Replaced.